File: prime-factors/PrimeFactorsApp-05/PrimeFactors/PrimeFactors.cs
namespace PrimeFactors;

public class PrimeFactors {
  readonly Sieve sieve;

  public PrimeFactors(Sieve sieve) {
    this.sieve = sieve;
  }

  public IEnumerable<long> Of(long n) {
    List<long> factors = new();
    foreach (long prime in sieve.GetPrimesUpTo(n)) {
      if (n > 1) {
        while (n % prime == 0) {
          factors.Add(prime);
          n /= prime;
        }
      }
    }

    if (n > 1) {
      factors.Add(n);
    }

    return factors;
  }
}

File: prime-factors/prime_factors_pyapp_07/prime_factors.py
def factors_of(n):
    factors = []
    for prime in [2]:
        while n % prime == 0:
            factors.append(prime)
            n //= prime

    if n > 1:
        factors.append(n)

    return factors


def primes_up_to(limit):
    if limit > 1:
        yield 2
